plugin/manager/manager.go

```diff
@@ -0,0 +1,240 @@
+// If the plugin does not have the required capability, the error is of type ErrNotCapable.
+// If the command starts but does not complete successfully, the error is of type RequestError wrapping a *exec.ExitError.
+// Other error types may be returned for other situations.
+func (mgr *Manager) Run(ctx context.Context, name string, cmd plugin.Command, args ...string) (interface{}, error) {
+	p, err := mgr.newPlugin(ctx, name)
+	if err != nil {
+		return nil, err
+	}
+	if p.Err != nil {
+		return nil, withErr(p.Err, ErrNotCompliant)
+	}
+	if c := cmd.Capability(); !p.HasCapability(c) {
+		return nil, ErrNotCapable
+	}
+	return run(ctx, mgr.cmder, p.Path, cmd, args...)
+}
+
+// newPlugin determines if the given candidate is valid and returns a Plugin.
+func (mgr *Manager) newPlugin(ctx context.Context, name string) (*Plugin, error) {
+	ok := isCandidate(mgr.fsys, name)
+	if !ok {
+		return nil, ErrNotFound
+	}
+
+	p := &Plugin{Path: binPath(mgr.fsys, name)}
+	out, err := run(ctx, mgr.cmder, p.Path, plugin.CommandGetMetadata)
+	if err != nil {
+		p.Err = fmt.Errorf("failed to fetch metadata: %w", err)
+		return p, nil
+	}
+	p.Metadata = *out.(*plugin.Metadata)
+	if p.Name != name {
+		p.Err = fmt.Errorf("executable name must be %q instead of %q", addExeSuffix(NamePrefix+p.Name), filepath.Base(p.Path))
+	} else if err := p.Metadata.Validate(); err != nil {
+		p.Err = fmt.Errorf("invalid metadata: %w", err)
+	}
+	return p, nil
+}
+
+// run executes the command and decodes the response.
+func run(ctx context.Context, cmder commander, pluginPath string, cmd plugin.Command, args ...string) (interface{}, error) {
+	out, err := cmder.Output(ctx, pluginPath, append([]string{string(cmd)}, args...)...)
+	if err != nil {
+		if ee, ok := err.(*exec.ExitError); ok {
+			var re plugin.RequestError
+			err = json.Unmarshal(ee.Stderr, &re)
+			if err == nil {
+				return nil, re
+			}
+			return nil, withErr(plugin.RequestError{Code: plugin.ErrorCodeGeneric, Err: ee}, ErrNotCompliant)
+		}
+		return nil, err
+	}
+	resp := cmd.NewResponse()
+	err = json.Unmarshal(out, resp)
+	if err != nil {
+		err = fmt.Errorf("failed to decode json response: %w", err)
+		return nil, withErr(err, ErrNotCompliant)
+	}
+	return resp, nil
+}
+
+// isCandidate checks if the named plugin is a valid candidate.
+func isCandidate(fsys fs.FS, name string) bool {
+	base := binName(name)
+	fi, err := fs.Stat(fsys, path.Join(name, base))
+	if err != nil {
+		// Ignore any file which we cannot Stat
+		// (e.g. due to permissions or anything else).
+		return false
+	}
+	if fi.Mode().Type() != 0 {
+		// Ignore non-regular files.
+		return false
+	}
+	return true
+}
+
+func binName(name string) string {
+	return addExeSuffix(NamePrefix + name)
+}
+
+func binPath(fsys fs.FS, name string) string {
+	base := binName(name)
+	if fsys, ok := fsys.(rootedFS); ok {
+		return filepath.Join(fsys.root, name, base)
+	}
+	return filepath.Join(name, base)
```

Comment:
I didn't understand what this code achieves, specially line 204. I'm new to Go.

Reply:
`if fsys, ok := fsys.(rootedFS); ok {` this is an interface type cast, being `ok` a flag which is true when `fsys` can be casted to `rootedFS`.

In production `fsys` type is always `rootedFS`, as this is what `NewManager` instantiates, but for testing purposes I'm also supporting generic `fsys` types, aka `fs.FS`, which are mostly instantiated from `testing/fstest` package.

Reply:
Got it. Should we add a code comment here to make this clearer, that the alternate path is to support testing?

Reply:
Done!